pkg/agent/packetcapture/packetcapture_controller_linux.go

```diff
@@ -0,0 +1,885 @@
+	}
+}
+
+func getPacketFileAndWriter(name string) (afero.File, *pcapgo.NgWriter, error) {
+	filePath := nameToPath(name)
+	var file afero.File
+	if _, err := os.Stat(filePath); err == nil {
+		return nil, nil, fmt.Errorf("the packet file %s already exists. This may be caused by an unexpected termination", filePath)
+	} else if os.IsNotExist(err) {
+		file, err = defaultFS.Create(filePath)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to create pcapng file: %w", err)
+		}
+	} else {
+		return nil, nil, fmt.Errorf("couldn't check if the file exists: %w", err)
+	}
+	writer, err := pcapgo.NewNgWriter(file, layers.LinkTypeEthernet)
+	if err != nil {
+		return nil, nil, fmt.Errorf("couldn't initialize a pcap writer: %w", err)
+	}
+	return file, writer, nil
+}
+
+// getTargetCaptureDevice is trying to locate the target device for packet capture. If the target Pod does not exist on the current Node,
+// the agent on this Node will not perform the capture.
+func (c *Controller) getTargetCaptureDevice(pc *crdv1alpha1.PacketCapture) *string {
+	var pod, ns string
+	if pc.Spec.Source.Pod != nil {
+		pod = pc.Spec.Source.Pod.Name
+		ns = pc.Spec.Source.Pod.Namespace
+	} else {
+		pod = pc.Spec.Destination.Pod.Name
+		ns = pc.Spec.Destination.Pod.Namespace
+	}
+
+	podInterfaces := c.interfaceStore.GetContainerInterfacesByPod(pod, ns)
+	if len(podInterfaces) == 0 {
+		return nil
+	}
+
+	return &podInterfaces[0].InterfaceName
+}
+
+func (c *Controller) startPacketCapture(name string) {
+	klog.V(4).InfoS("Started processing PacketCapture", "name", name)
+
+	go func() error {
+		var err error
+		pcState := &packetCaptureState{name: name}
+		defer func() {
+			if err != nil {
+				c.cleanupPacketCapture(name)
+				klog.ErrorS(err, "PackageCapture failed", "name", name)
+				err := c.updatePacketCaptureStatus(name, pcState.capturedPacketsNum, "", err)
+				if err != nil {
+					klog.ErrorS(err, "failed to update PacketCapture status")
+				}
+			}
+		}()
+		pc, err := c.packetCaptureLister.Get(name)
+		if err != nil {
+			return err
+		}
+
+		device := c.getTargetCaptureDevice(pc)
+		if device == nil {
+			return nil
+		}
+
+		klog.V(2).InfoS("Prepare capture on current node", "name", pc.Name, "device", *device)
+		packetSource, err := c.newPacketSourceFn(c, pc)
+		if err != nil {
+			klog.ErrorS(err, "failed to create packet source")
+			return err
+		}
+		if packetSource == nil {
+			return nil
+		}
+
+		pcState.targetCapturedPacketsNum = pc.Spec.CaptureConfig.FirstN.Number
+		file, writer, err := getPacketFileAndWriter(pc.Name)
+		if err != nil {
+			return err
+		}
+
+		pcState.pcapngFile = file
+		pcState.pcapngWriter = writer
+		pcState.updateRateLimiter = rate.NewLimiter(rate.Every(captureStatusUpdatePeriod), 1)
+		timeout := defaultTimeoutDuration
+		if pc.Spec.Timeout != nil {
+			timeout = time.Duration(*pc.Spec.Timeout) * time.Second
+		}
+
+		err = c.performCapture(pcState, timeout, packetSource)
```

Comment:
up until this point, the code can be synchronous (not execute in a separate goroutine). Only the capture itself needs to be asynchronous.

Reply:
+1, including the defered function. Otherwise it would be really difficult to figure out the reponsibilities and potential race condition between the worker's goroutine and the capturer's goroutine.

Ideally the capture only capture packets and notify the workers to maintain K8s object status by enqueuing the capture name once it succeeds or fails.

Reply:
updated.  

> Ideally the capture only capture packets and notify the workers to maintain K8s object status by enqueuing the capture name once it succeeds or fails.

currently  the code will sync the CR status directly in the go routine. Does that works?